From 595341eb7895b233bb2301b7cb90799c9f13a8d8 Mon Sep 17 00:00:00 2001
From: Venkatarama Avadhani <venkataramaa@google.com>
Date: Wed, 15 Apr 2026 11:45:51 +0530
Subject: [PATCH 1/8] Fix heap-buffer-overflow in AudioAttributes unmarshalling

This CL fixes a heap-buffer-overflow in MediaPlayerService's AudioAttributes
unmarshalling logic. The source length was in bytes and the
utf16_to_utf8 uses char16_t pointer arithmetic on it that would lead to
an OOB read. Corrected the source length parameter to give length as
number of char16_t.
The destination length has to be 1 less than the actual size of the
destination since utf16_to_utf8 adds a NULL terminator. Correct this
behaviour to limit the destination length to allocated length - 1.

Bug: 482172329
Bug: 396471524

Test: fuzzer with the testcases
Test: atest CtsMediaAudioTestCases CtsMediaPlayerTestCases libutils_tests

Flag: EXEMPT CVE_FIX
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:50d6c98e00c2c5620ea50e26c9c44e70a0c5ce8a
Merged-In: I9f607d57281c76d8e1c1a14112dffdf22acb8080
Change-Id: I9f607d57281c76d8e1c1a14112dffdf22acb8080
---
 media/libmediaplayerservice/MediaPlayerService.cpp | 7 ++-----
 1 file changed, 2 insertions(+), 5 deletions(-)

diff --git a/media/libmediaplayerservice/MediaPlayerService.cpp b/media/libmediaplayerservice/MediaPlayerService.cpp
index b8ff92f2c1..cb28c46959 100644
--- a/media/libmediaplayerservice/MediaPlayerService.cpp
+++ b/media/libmediaplayerservice/MediaPlayerService.cpp
@@ -243,15 +243,12 @@ void unmarshallAudioAttributes(const Parcel& parcel, audio_attributes_t *attribu
         // the tags are UTF16, convert to UTF8
         String16 tags = parcel.readString16();
         ssize_t realTagSize = utf16_to_utf8_length(tags.c_str(), tags.size());
-        if (realTagSize <= 0) {
+        if (realTagSize <= 0 || realTagSize >= AUDIO_ATTRIBUTES_TAGS_MAX_SIZE) {
             strcpy(attributes->tags, "");
         } else {
             // copy the flattened string into the attributes as the destination for the conversion:
             // copying array size -1, array for tags was calloc'd, no need to NULL-terminate it
-            size_t tagSize = realTagSize > AUDIO_ATTRIBUTES_TAGS_MAX_SIZE - 1 ?
-                    AUDIO_ATTRIBUTES_TAGS_MAX_SIZE - 1 : realTagSize;
-            utf16_to_utf8(tags.c_str(), tagSize, attributes->tags,
-                    sizeof(attributes->tags) / sizeof(attributes->tags[0]));
+            utf16_to_utf8(tags.c_str(), tags.size(), attributes->tags, realTagSize);
         }
     } else {
         ALOGE("unmarshallAudioAttributes() received unflattened tags, ignoring tag values");
-- 
2.55.0

